smart-py/objkt_swap_v3.py

```diff
@@ -0,0 +1,363 @@
+
+This version corrects several small bugs from the v2 marketplace contract and
+adds the possibility to trade different kinds of FA2 tokens.
+
+"""
+
+import smartpy as sp
+
+
+class Marketplace(sp.Contract):
+    """This contract implements the next version of the H=N marketplace
+    contract.
+
+    """
+
+    SWAP_TYPE = sp.TRecord(
+        # The user that created the swap
+        issuer=sp.TAddress,
+        # The token FA2 contract address
+        fa2=sp.TAddress,
+        # The token id (not necessarily from a OBJKT)
+        objkt_id=sp.TNat,
+        # The number of swapped editions
+        objkt_amount=sp.TNat,
+        # The price of each edition in mutez
+        xtz_per_objkt=sp.TMutez,
+        # The artists royalties in (1000 is 100%)
+        royalties=sp.TNat,
+        # The address that will receive the royalties
+        creator=sp.TAddress).layout(
+            ("issuer", ("fa2", ("objkt_id", ("objkt_amount", ("xtz_per_objkt", ("royalties", "creator")))))))
+
+    def __init__(self, manager, metadata, allowed_fa2s, fee):
+        """Initializes the contract.
+
+        """
+        # Define the contract storage data types for clarity
+        self.init_type(sp.TRecord(
+            manager=sp.TAddress,
+            fee_recipient=sp.TAddress,
+            metadata=sp.TBigMap(sp.TString, sp.TBytes),
+            allowed_fa2s=sp.TBigMap(sp.TAddress, sp.TBool),
+            swaps=sp.TBigMap(sp.TNat, Marketplace.SWAP_TYPE),
+            fee=sp.TNat,
+            counter=sp.TNat,
+            paused=sp.TBool))
+
+        # Initialize the contract storage
+        self.init(
+            manager=manager,
+            fee_recipient=manager,
+            metadata=metadata,
+            allowed_fa2s=allowed_fa2s,
+            swaps=sp.big_map(),
+            fee=fee,
+            counter=0,
+            paused=False)
+
+    def check_is_manager(self):
+        """Checks that the address that called the entry point is the contract
+        manager.
+
+        """
+        sp.verify(sp.sender == self.data.manager,
+                  message="This can only be executed by the manager")
+
+    def check_is_not_paused(self):
+        """Checks that the contract is not paused.
+
+        """
+        sp.verify(~self.data.paused,
+                  message="This contract is paused")
+
+    def check_no_tez_transfer(self):
+        """Checks that no tez were transferred in the operation.
+
+        """
+        sp.verify(sp.amount == sp.tez(0),
+                  message="The operation does not need tez transfers")
+
+    @sp.entry_point
+    def swap(self, params):
+        """Swaps several editions of a token for a fixed price.
+
+        """
+        # Define the input parameter data type
+        sp.set_type(params, sp.TRecord(
+            fa2=sp.TAddress,
+            objkt_id=sp.TNat,
+            objkt_amount=sp.TNat,
+            xtz_per_objkt=sp.TMutez,
+            royalties=sp.TNat,
```

Comment:
the royalties are stored on the swap? is this not built in to the objkt token?

Reply:
Unfortunately we have the same limitations that the v2 contract had. Royalties info is in the v1 marquetplace contract and cannot be accessed during the v2 or v3 contract execution.
We will have to rely on the same UI implementation that we have before:
- when the user clicks the swap button, they can only fill the editions and the price per edition. The fa2 address, the artist, and the royalties will be filled by the UI with the info provided by hicdex
- if someone makes the swap using better-call.dev, the swap will be visible in the H=N site only if the creator and the royalties coincide. Otherwise, it will not be visible at the site and people will not be able to collect the swap. Of course, people could still collect the swap using better-call.dev

Reply:
We should change that if we mint another FA2 token (newOBJKT). In that case we could save the royalties in the token contract itself and create a view that is easily accessible by marketplace contracts.
https://smartpy.io/releases/20210917-545ba21ad167a0c2d9f7e99ce41a5015876cf62e/docs/experimental/onchain_views/